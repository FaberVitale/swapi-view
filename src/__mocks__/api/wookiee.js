export default {
  whrascwo: "Yrahoahwh IV",
  rcooaoraaoahoowh_akworcahoowa: "24",
  oorcrhahaoraan_akworcahoowa: "4818",
  waahrascwoaoworc: "10200",
  oaanahscraaowo: "aowoscakworcraaowo, aorcooakahoaraan",
  rrrcrahoahaoro: "1 caorawhwararcwa",
  aoworcrcraahwh: "shhuwhrranwo, rcraahwhwwoorcwocaoc",
  churcwwraoawo_ohraaoworc: "8",
  akooakhuanraaoahoowh: "1000",
  rcwocahwawowhaoc: [],
  wwahanscc: ["acaoaoakc://cohraakah.oaoo/raakah/wwahanscc/1/"],
  oarcworaaowowa: "2014-12-10T11:37:19.144000Z",
  wowaahaowowa: "2014-12-20T20:58:18.421000Z",
  hurcan: "acaoaoakc://cohraakah.oaoo/raakah/akanrawhwoaoc/3/"
};
