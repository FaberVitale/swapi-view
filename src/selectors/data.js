//@flow
import type { State } from "../reducers";
import type { Match } from "react-router-dom";
import { createSelector } from "reselect";
import { endpoints, searchFieldsOf } from "../routes";

export const getData = (state: State) => state.data;

const getEndpointData = (state: State, match: Match) => {
  const data = getData(state);
  const endpoint = match.params.endpoint || null;

  return data && endpoint ? data[endpoint] : null;
};

export const hasResource = (state: State, request: FetchRequest) => {
  const { match, type, query, name } = request;

  switch (type) {
    case "search":
      return !!getSearchPage(state, name, query);
    case "item":
      return !!getItem(state, match);
    case "page":
      return !!getPage(state, match);
    case "wookiee":
      return !!getWookiee(state, match);
    default:
      return false;
  }
};

const getItem = (state: State, match: Match) => {
  const endpointData = getEndpointData(state, match);

  return endpointData && endpointData.items[match.url];
};

export const getPage = (state: State, match: Match) => {
  const endpointData = getEndpointData(state, match);

  return endpointData && endpointData.pages[match.url];
};

const getWookiee = (state: State, match: Match) => {
  const endpointData = getEndpointData(state, match);

  return endpointData && endpointData.wookiee[match.url];
};

const defaultSearchPage = {
  previous: null,
  next: null,
  items: []
};

export const getSearchPage = (
  state: State,
  name: string,
  query: { q: string, page: string } | null
) => {
  if (!query) {
    return defaultSearchPage;
  }

  return getData(state).search[name] || null;
};

export const getItemOrWookiee = (state: State, match: Match) => {
  return match.params.type === "item"
    ? getItem(state, match)
    : getWookiee(state, match);
};

/* getItemsOf[endpoint]:  (data) => data[endpoint].items
 * in order to work, it requires that the input is the result of getData
 * */
export const getItemsOf = endpoints.reduce((res, endpoint) => {
  res[endpoint] = data => data[endpoint].items;

  return res;
}, {});

/* Object that maps endpoints to  a function that creates an array of suggestions
 * given the endpoint given the result of getData  */
export const getSearchSuggestionsOf = (function buildSearchSuggestionsOf(
  endpoints
) {
  const retrieveSuggestions = (endpoint: string) => items => {
    let suggestions: Array<string> = [];
    const keys: Array<string> = Object.keys(items);

    const suggestionToKey = {};

    for (const key of keys) {
      const searchValues = searchFieldsOf[endpoint].map(
        searchValue => items[key].data[searchValue]
      );
      suggestions.push(...searchValues);

      for (const searchValue of searchValues) {
        suggestionToKey[searchValue] = key;
      }
    }

    // a lot vehicles and starships have duplicated values
    if (suggestions.length > 1 && searchFieldsOf[endpoint].length > 1) {
      suggestions = [...new Set(suggestions)];
    }

    return {
      suggestions,
      suggestionToKey
    };
  };

  let res: {
    [endpoint: string]: (items: {}) => {
      suggestion: Array<string>,
      suggestionToKey: { [sugg: string]: string }
    }
  } = {};

  for (const endpoint of endpoints) {
    res[endpoint] = createSelector(
      getItemsOf[endpoint],
      retrieveSuggestions(endpoint)
    );
  }

  return res;
})(endpoints || []);

/* given (state: State) passed returns an object that maps endoints to suggestions list */
export const getDataSources = createSelector(getData, function buildDatasources(
  data
) {
  const res = {};

  for (let i = 0, len = endpoints.length; i < len; i++) {
    res[endpoints[i]] = getSearchSuggestionsOf[endpoints[i]](data);
  }

  return res;
});
