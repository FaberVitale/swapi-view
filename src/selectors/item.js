//@flow
const defaultItem = {
  data: {},
  meta: {},
  links: {}
};

export const getEditedKey = (item: ItemData = defaultItem) => {
  const meta = item.meta;

  if (!meta) {
    return "";
  }

  if (typeof meta.edited === "string") {
    return "edited";
  }

  if (typeof meta.wowaahaowowa === "string") {
    return "wowaahaowowa";
  }
  return "";
};

export const getItemName = (item: ItemData = defaultItem) => {
  const data = item.data;

  if (!data) {
    return "";
  }

  return data.name || data.title || data.whrascwo || "";
};

export const getItemNamesOfPage: (
  page: PageData,
  items: { [route: string]: ItemData }
) => Array<string> = (page, items) => {
  const routes = page.items;

  return routes.map(itemRoute => getItemName(items[itemRoute]));
};
