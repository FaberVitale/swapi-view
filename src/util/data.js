//@flow
const isMetaKey = (key: string) =>
  /^(?:created|url|edited|oarcworaaowowa|wowaahaowowa|hurcan)$/.test(key);
const isLink = value =>
  typeof value !== "string" || /^https?|acaoaoakc/.test(value);

const regex = /(\d+)\/?$/;
export const mapTypeToRoute = (
  type: string,
  endpoint: string,
  apiUrl: string,
  query?: { q: string, page: string } | null
) => {
  const res = regex.exec(apiUrl);

  if (type === "search" && query && res) {
    return `/${endpoint}/search/${query.q}/${res[1]}`;
  }

  return res && `/${endpoint}/${type}/${res[1]}`;
};

export const transformItem = (item: {}) =>
  Object.keys(item || {}).reduce(
    (aggr, key) => {
      const value = item[key];

      if (isMetaKey(key)) {
        aggr.meta[key] = value;
      } else if (isLink(value)) {
        aggr.links[key] = value;
      } else {
        aggr.data[key] = value;
      }
      return aggr;
    },
    { meta: {}, data: {}, links: {} }
  );

export const transformPage = (
  request: FetchRequest,
  page: {
    next: string | null,
    previous: string | null,
    results: Array<{ url: string }>
  }
) => {
  const type = request.type;
  const { endpoint } = request.match.params;

  return {
    next: page.next && mapTypeToRoute(type, endpoint, page.next, request.query),
    previous:
      page.previous &&
      mapTypeToRoute(type, endpoint, page.previous, request.query),
    ids: page.results.map(item => mapTypeToRoute("item", endpoint, item.url)),
    results: page.results.map(transformItem)
  };
};
