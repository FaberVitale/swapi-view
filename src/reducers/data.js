//@flow
import { equalsTo, equals } from "../util/functions";
import { combineReducers, compose } from "redux";
import { filterActions } from "redux-ignore";
import pagination from "./pagination";
import byId from "./byId";
import type { State as ById } from "./byId";
import type { State as PageState } from "./pagination";
import { endpoints } from "../routes";
import { getType, getEndpoint } from "../selectors/router";

type EndpointData = {|
  wookiee: ById,
  items: ById,
  pages: PageState
|};

type MakeEndpointReducer = string => (
  state: EndpointData,
  action: Action
) => EndpointData;

type MakeDataReducer = (
  endpoints: Array<string>
) => {
  [endpoint: string]: (state: EndpointData, action: Action) => EndpointData
};

export type State = {
  [endpoint: string]: EndpointData,
  search: PageState
};

export default (function buildDataReducer(endpoints) {
  const filterFor = {
    page: compose(equals("page"), getType),
    item: compose(equalsTo("page", "item", "search"), getType),
    wookiee: compose(equals("wookiee"), getType),
    search: compose(equals("search"), getType)
  };

  const makeEndpointReducer: MakeEndpointReducer = (endpoint: string) =>
    filterActions(
      combineReducers({
        pages: filterActions(pagination, filterFor.page),
        items: filterActions(byId, filterFor.item),
        wookiee: filterActions(byId, filterFor.wookiee)
      }),
      compose(equals(endpoint), getEndpoint)
    );

  const makeDataReducer: MakeDataReducer = endpoints => {
    const res = endpoints.reduce((aggr, next) => {
      aggr[next] = makeEndpointReducer(next);

      return aggr;
    }, {});

    res.search = filterActions(pagination, filterFor.search);

    return res;
  };

  return combineReducers(makeDataReducer(endpoints));
})(endpoints || []);
