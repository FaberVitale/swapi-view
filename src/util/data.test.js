import * as data from "./data";
import page from "../__mocks__/api/page";
import item from "../__mocks__/api/item";
import wookiee from "../__mocks__/api/wookiee";

describe("util/data", () => {
  test("transformItem", () => {
    expect(data.transformItem(item)).toMatchSnapshot();
    expect(data.transformItem(wookiee)).toMatchSnapshot();
  });

  test("transformPage", () => {
    expect("transformPage", () => {
      expect(data.transformItem(page)).toMatchSnapshot();
    });
  });
});
