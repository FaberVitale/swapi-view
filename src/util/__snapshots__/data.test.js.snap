// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`util/data transformItem 1`] = `
Object {
  "data": Object {
    "birth_year": "19BBY",
    "eye_color": "blue",
    "gender": "male",
    "hair_color": "blond",
    "height": "172",
    "mass": "77",
    "name": "Luke Skywalker",
    "skin_color": "fair",
  },
  "links": Object {
    "films": Array [
      "https://swapi.co/api/films/2/",
      "https://swapi.co/api/films/6/",
      "https://swapi.co/api/films/3/",
      "https://swapi.co/api/films/1/",
      "https://swapi.co/api/films/7/",
    ],
    "homeworld": "https://swapi.co/api/planets/1/",
    "species": Array [
      "https://swapi.co/api/species/1/",
    ],
    "starships": Array [
      "https://swapi.co/api/starships/12/",
      "https://swapi.co/api/starships/22/",
    ],
    "vehicles": Array [
      "https://swapi.co/api/vehicles/14/",
      "https://swapi.co/api/vehicles/30/",
    ],
  },
  "meta": Object {
    "created": "2014-12-09T13:50:51.644000Z",
    "edited": "2014-12-20T21:17:56.891000Z",
    "url": "https://swapi.co/api/people/1/",
  },
}
`;

exports[`util/data transformItem 2`] = `
Object {
  "data": Object {
    "akooakhuanraaoahoowh": "1000",
    "aoworcrcraahwh": "shhuwhrranwo, rcraahwhwwoorcwocaoc",
    "churcwwraoawo_ohraaoworc": "8",
    "oaanahscraaowo": "aowoscakworcraaowo, aorcooakahoaraan",
    "oorcrhahaoraan_akworcahoowa": "4818",
    "rcooaoraaoahoowh_akworcahoowa": "24",
    "rrrcrahoahaoro": "1 caorawhwararcwa",
    "waahrascwoaoworc": "10200",
    "whrascwo": "Yrahoahwh IV",
  },
  "links": Object {
    "rcwocahwawowhaoc": Array [],
    "wwahanscc": Array [
      "acaoaoakc://cohraakah.oaoo/raakah/wwahanscc/1/",
    ],
  },
  "meta": Object {
    "hurcan": "acaoaoakc://cohraakah.oaoo/raakah/akanrawhwoaoc/3/",
    "oarcworaaowowa": "2014-12-10T11:37:19.144000Z",
    "wowaahaowowa": "2014-12-20T20:58:18.421000Z",
  },
}
`;
